
This directory contains simulated fastq reads and truth VCFs for insertions, deletions, MNVs, and tandem duplications.

The fastqs contain paired-end, 100bp Illumina-style reads, are free of basecalling errors, and should produce alignments that
cover each variant to a total depth of approximately 200 or 30.

Each directory contains truth VCFs and fastqs for four separate variant types: insertions of random sequence ("ins"), deletions ("dels"), tandem duplications ("dups"), and length-preserving
MNVs ("mnps").  For each variant type, 10 instances of each size in (1,2,3,4,5, 10, 20, 30... 150) were randomly generated by selecting a position from all 
exonic regions, with no padding into intronic regions. 

When comparing the truth VCF to variant caller output, it is highly recommended to use a graph-based variant comparison tool such as [vgraph](https://github.com/bioinformed/vgraph),
[vcfeval](https://github.com/RealTimeGenomics/rtg-tools), or [hap.py](https://github.com/Illumina/hap.py) to match haplotypes. Other normalization/ comparison methods such as /vt/, /vcfellelicprimitives/, etc., may not correctly match all variants, especially in the case of MNPs.  
